### DISCUSSION TITLE
Manual Jobs

### DISCUSSION THREAD

**Discussion by brpaz** [community/community#25274] [Actions]
<p>It would be great if GitHub actions could support running jobs manually.</p>
<p>A very common use case for this, is for release or deploy to production envrionment where you want to control when it will be done or do some manual checks before (Ex: QA or Client validation in staging environment).</p>
<p>GitLab supports this and it could be done by having a new “manual”: key on the workflow file. When set to “true”, a button would appear to the interface to trigger the job.</p>

**Comment by tedmiston** [comment, marked as answer]
<p>The workaround I use for manual workflows is to set the trigger to the repository_dispatch webhook event, and then send a POST request to it locally.</p>
<p>GitHub Actions setup - <a href="https://help.github.com/en/articles/events-that-trigger-workflows#external-events" rel="nofollow noopener">https://help.github.com/en/articles/events-that-trigger-workflows#external-events</a></p>
<p>Calling it - <a href="https://developer.github.com/v3/repos/#create-a-repository-dispatch-event" rel="nofollow noopener">https://developer.github.com/v3/repos/#create-a-repository-dispatch-event</a></p>
<p>This also works well if you want to programmatically trigger a GitHub Actions workflow from another system.</p>

**Comment by ansman** [comment]
<p>Unfortunately this only allows you to trigger the build on the default branch, not on custom branches.</p>

**Comment by tedmiston** [comment]
<p>I’m not sure if this works for your use case, but I think you could have the workflow be triggered on the default branch, but then contain a checkout action that checks out your desired non-default branch, etc.</p>
<pre><code>-uses:actions/checkoutwith:
ref:my-branch
</code></pre>
<p><a href="https://github.com/actions/checkout/blob/f6ce2afa7079cb075a124c93c79d61779d845782/action.yml#L6-L7" rel="nofollow noopener">https://github.com/actions/checkout/blob/f6ce2afa7079cb075a124c93c79d61779d845782/action.yml#L6-L7</a></p>
<p>I have not tested it personally, but I’m thinking that using this combined with the path arg, that you could then check out multiple copies of your repo as desired to do processing on the code from several different non-default branches.  Does this work for your use case?</p>

**Comment by abatilo** [comment]
<p>Hi all, I’m not entirely sure if this is appropriate or not to self promote here, but my buddy and I built an extremely basic MVP for handling this exact situation, because we wanted this feature as well:</p>
<p><a href="https://www.actionspanel.app" rel="nofollow noopener">https://www.actionspanel.app</a></p>
<p>We built it as fast as we could so it’s still very quirky, but we’d love feedback. There’s a contact email on that landing page.</p>

**Comment by peter-evans** [comment]
<p>Another option is a slash command style “ChatOps” solution.</p>
<p><a href="https://github.com/peter-evans/slash-command-dispatch" rel="nofollow noopener">https://github.com/peter-evans/slash-command-dispatch</a></p>

**Comment by jk2K** [comment]
<p>now, You can approve or reject jobs awaiting review.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments">
  <header class="source">
      <img src="https://docs.github.com/assets/cb-803/images/site/favicon.svg" class="site-icon" width="16" height="16">

      <a href="https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments" target="_blank" rel="noopener nofollow ugc">GitHub Docs</a>
  </header>

  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/4025839/181090041-50a59f56-2eee-4c19-b84c-6a74a8a0dc2a.png" class="thumbnail onebox-avatar" width="500" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_500x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_750x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_1000x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/3/337b7cc06184984e148c075df20ef009e0e6ecaa_2_10x10.png">

<h3><a href="https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments" target="_blank" rel="noopener nofollow ugc">Reviewing deployments - GitHub Docs</a></h3>

  <p>You can approve or reject jobs awaiting review.</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Reply by osher** [reply]
Well. My usecase is a little more demanding. 
Consider a "monorepo" of terraform deployments.
the repo has the following structure:
```
+ deployments
   + product1
      + env1
         + _tf-backend-state.tf
         + resource-set1.tf
         + resource-set1.tf
      + env2...
   + product2...
+ modules
   + resource-set1...
   + resource-set2...
```
Given that terraform is not that good at detecting dependencies, user supervision is advised.

However, we want to keep the user supervision to the decision level, and encapsulate all the details in jobs.

## requirements:
  - target detection
    - when a deployment is updated - it should be run
    - when a module is updated - all deployments in which it is involved should be run
    - a target should not run twice
  - execution
    - change in a PR should run only `terraform plan` for all detected targets
    - changes in main branch should let the user decide to run `terraform apply` for each target at their own discretion

This last point is ...hard here with github actions.

## competitor product: gitlab-ci
On gitlab CI I have on the base workflow a code-generator that creates `ci.yaml` dynamically, which includes only the jobs for the detected targets as implied by the observed changes (or filters on jobs that effectively run only the relevant jobs).

Then the dynamically triggered child-pipeline with the generated content runs `terraform plan` for all detected targets, and - for the main branch - contains jobs with the `terraform apply` command that the user can fire manually at their own discretion, each depends on its respective plan job.

I wish I could do the same here.


## UPDATED
somebody threw me a direction of using a matrix, where the spec of the matrix is composed as an output of a preceding job in the form of:
```
strategy:
      matrix:
        TARGET: "${{ fromJSON(needs.changes.outputs.apps) }}"
```
where `changes` is the preceding job

That could be interesting because it will run deployments in parallel, however, it still lacks the ability to supervise the order between them when it's needed. I need to think about it more :thinking:
